# Ryan's Titanic Dataset Exploration
**Author:** Ryan Krabbe  
**Date:** 4/02/2025

**Objective:** Use the Titantic dataset to explore different machine learning models to predict passenger survival.


## Introduction
This project uses the Titantic dataset to explore different features that led to survival patterns such as age, sex, gender and fare. In this project, I'll incorporate different machine models and techniques to predict passenger survival.

## Imports
In the code cell below, import the necessary Python libraries for this notebook.  

In [141]:
import seaborn as sns
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression, Ridge, ElasticNet
from sklearn.preprocessing import PolynomialFeatures
from sklearn.metrics import mean_squared_error, mean_absolute_error, r2_score

## Section 1. Import and Inspect the Data

### Load the Titanic dataset and confirm it’s structured correctly

In [142]:
# Load Titanic dataset from seaborn and verify
titanic = sns.load_dataset("titanic")
titanic.head()

survived  pclass     sex   age  sibsp  parch     fare embarked  class  \
0         0       3    male  22.0      1      0   7.2500        S  Third   
1         1       1  female  38.0      1      0  71.2833        C  First   
2         1       3  female  26.0      0      0   7.9250        S  Third   
3         1       1  female  35.0      1      0  53.1000        S  First   
4         0       3    male  35.0      0      0   8.0500        S  Third   

     who  adult_male deck  embark_town alive  alone  
0    man        True  NaN  Southampton    no  False  
1  woman       False    C    Cherbourg   yes  False  
2  woman       False  NaN  Southampton   yes   True  
3  woman       False    C  Southampton   yes  False  
4    man        True  NaN  Southampton    no   True

## Section 2. Data Exploration and Preparation

### Load the Titanic dataset and confirm it’s structured correctly

In [143]:
# Fill missing 'age' with median
titanic['age'] = titanic['age'].fillna(titanic['age'].median())

# Fill missing 'sex' with mode before encoding
titanic['sex'] = titanic['sex'].fillna(titanic['sex'].mode().iloc[0])
titanic['sex'] = titanic['sex'].map({'male': 0, 'female': 1})

# Drop rows with missing fare
titanic = titanic.dropna(subset=['fare'])

# Create family_size
titanic['family_size'] = titanic['sibsp'] + titanic['parch'] + 1

In [145]:
# Case 1. age only
X1 = titanic[['age']]
y1 = titanic['fare']

# Case 2. family_size only
X2 = titanic[['family_size']]
y2 = titanic['fare']

# Case 3. age and family size
X3 = titanic[['age', 'family_size']]
y3 = titanic['fare']

# Case 4. age, pclass, sex and embarked
X4 = titanic[['age', 'pclass', 'sex']]
y4 = titanic['fare']

## Section 3. Feature Selection and Justification

Reflection Questions

Why might these features affect a passenger’s fare: 
    - `age`: Children might have different fares compared to adults
    - `family_size`: Families might book together, which could result in reduced passenger fares
    - `sex`: Sex could interact with other features such as `family_size` or `embarked`
    - `pclass` - This is strongly correlated with fare based on the class that passengers were in
List all available features:
- `survived`
- `pclass`
- `sex`
- `age`
- `sibsp`
- `parch`
- `fare`
- `embarked`
- `class`
- `who`
- `adult_male`
- `deck`
- `embark_town`
- `alive`
- `alone`
Which other features could improve predictions and why:
- `alone`: If a passenger is traveling alone it could lead to different prices within the fare
- `ticket`: Possibly the ticket could provide minor details about a passengers fare
How many variables are in your Case 4:
- 3 variables
Which variable(s) did you choose for Case 4 and why do you feel those could make good inputs: 
- `age`: Children might have different fares compared to adults
- `sex`: Sex could interact with other features such as `family_size` or `embarked`
- `pclass` - This is strongly correlated with fare based on the class that passengers were in

## Section 4. Train a Regression Model (Linear Regression)

## 4.1 Split the Data

In [148]:
X1_train, X1_test, y1_train, y1_test = train_test_split(X1, y1, test_size=0.2, random_state=123)

X2_train, X2_test, y2_train, y2_test = train_test_split(X2, y2, test_size=0.2, random_state=123)

X3_train, X3_test, y3_train, y3_test = train_test_split(X3, y3, test_size=0.2, random_state=123)

X4_train, X4_test, y4_train, y4_test = train_test_split(X4, y4, test_size=0.2, random_state=123)

## 4.2 Train and Evaluate Linear Regression Models (all 4 cases)

In [150]:
lr_model1 = LinearRegression().fit(X1_train, y1_train)
lr_model2 = LinearRegression().fit(X2_train, y2_train)
lr_model3 = LinearRegression().fit(X3_train, y3_train)
lr_model4 = LinearRegression().fit(X4_train, y4_train)

# Predictions

y_pred_train1 = lr_model1.predict(X1_train)
y_pred_test1 = lr_model1.predict(X1_test)

y_pred_train2 = lr_model2.predict(X2_train)
y_pred_test2 = lr_model2.predict(X2_test)

y_pred_train3 = lr_model3.predict(X3_train)
y_pred_test3 = lr_model3.predict(X3_test)

y_pred_train4 = lr_model4.predict(X4_train)
y_pred_test4 = lr_model4.predict(X4_test)

## 4.3 Report Performance

In [156]:
from sklearn.metrics import r2_score, mean_squared_error, mean_absolute_error

print("Case 1: Training R²:", r2_score(y1_train, y_pred_train1))
print("Case 1: Test R²:", r2_score(y1_test, y_pred_test1))
print("Case 1: Test RMSE:", mean_squared_error(y1_test, y_pred_test1) ** 0.5)
print("Case 1: Test MAE:", mean_absolute_error(y1_test, y_pred_test1))

print("Case 2: Training R²:", r2_score(y2_train, y_pred_train2))
print("Case 2: Test R²:", r2_score(y2_test, y_pred_test2))
print("Case 2: Test RMSE:", mean_squared_error(y2_test, y_pred_test2) ** 0.5)
print("Case 2: Test MAE:", mean_absolute_error(y2_test, y_pred_test2))

print("Case 3: Training R²:", r2_score(y3_train, y_pred_train3))
print("Case 3: Test R²:", r2_score(y3_test, y_pred_test3))
print("Case 3: Test RMSE:", mean_squared_error(y3_test, y_pred_test3) ** 0.5)
print("Case 3: Test MAE:", mean_absolute_error(y3_test, y_pred_test3))

print("Case 4: Training R²:", r2_score(y4_train, y_pred_train4))
print("Case 4: Test R²:", r2_score(y4_test, y_pred_test4))
print("Case 4: Test RMSE:", mean_squared_error(y4_test, y_pred_test4) ** 0.5)
print("Case 4: Test MAE:", mean_absolute_error(y4_test, y_pred_test4))

Case 1: Training R²: 0.009950688019452314
Case 1: Test R²: 0.0034163395508415295
Case 1: Test RMSE: 37.97164180172938
Case 1: Test MAE: 25.28637293162364
Case 2: Training R²: 0.049915792364760736
Case 2: Test R²: 0.022231186110131973
Case 2: Test RMSE: 37.6114940041967
Case 2: Test MAE: 25.02534815941641
Case 3: Training R²: 0.07347466201590014
Case 3: Test R²: 0.049784832763073106
Case 3: Test RMSE: 37.0777586646559
Case 3: Test MAE: 24.284935030470688
Case 4: Training R²: 0.3151329425803896
Case 4: Test R²: 0.35025268636237106
Case 4: Test RMSE: 30.66018284544514
Case 4: Test MAE: 20.47729671828975


### Section 4 Reflection Questions

Compare the train vs test results for each.

Did Case 1 overfit or underfit? Explain: Underfit meaning age is not predictive of fare.
Did Case 2 overfit or underfit? Explain: Underfit, but not as much as case 1. Family_size is not very predictive of fare.
Did Case 3 overfit or underfit? Explain: Underfit, another improvement, but still not predictive of fare.
Did Case 4 overfit or underfit? Explain: The training R² of 0.32 and Test R² of 0.35 means this was a good fit and not underfit or overfit. These features were highly predictive of fare.

Adding Age

Did adding age improve the model: Yes, but very slightly. Age on it's own did not improve the model it only improved the model slighly when it was used with other features.
Propose a possible explanation (consider how age might affect ticket price, and whether the data supports that): Age is not indictive of fare like some other features are. For example, a 60 year old passenger who is traveling in third class would pay less than a 5 year old passenger in first class.

Worst

Which case performed the worst: Case 1
How do you know: Case 1 had the lowest R² and highest RMSE and MAE.
Do you think adding more training data would improve it (and why/why not): Maybe slightly, but honestly not too much. What I gathered from all of this is that age is not a strong predictor of fare.

Best

Which case performed the best: Case 4
How do you know: Case 4 had the highest R² and lowest RMSE and MAE.
Do you think adding more training data would improve it (and why/why not): Yes, I think so. The model already ran well with the features selected, but adding another strong feature could improve the model even more so.

## Section 5. Compare Alternative Models